src/components/BrowserCell/BrowserCell.react.js

```diff
@@ -23,12 +23,149 @@ export default class BrowserCell extends Component {
     this.cellRef = React.createRef();
     this.copyableValue = undefined;
     this.state = {
-      showTooltip: false
-    }
-
+      showTooltip: false,
+      content: null,
+      classes: []
+    };
   }
 
-  componentDidUpdate(prevProps) {
+  async componentDidUpdate(prevProps) {
```

Comment:
What are these changes? They seem not to be part of this PR?

Reply:
The value of pointer key column is fetched from local storage using await, that's why async is added here.
https://github.com/fn-faisal/parse-dashboard/blob/feat/151-pcd/src/components/BrowserCell/BrowserCell.react.js#L376

Reply:
I actually mean the many changes in this method:

https://github.com/parse-community/parse-dashboard/pull/1723/files?file-filters%5B%5D=.js#diff-4e4d2ae32c4431bc66b52e869432797428508c6a3cb0811018e658525f0637c9R33-R168

Reply:
I moved the changes from the component's render method to the did update method. I was having issues with async loading of data ( like key from local storage ).

Reply:
When you say "moved", I don't see nearly the same amount of lines removed from the `render` method. That looks strange to me.

Reply:
Oh, sorry, I copied cell rendering logic to re-render on component updates for async calls. But this is for component updates. The lines in the render are for the initial render.

Reply:
Do you mean that you copied these lines from somewhere else? From where?

Reply:
The lines are what render the browsercell based on the type of data passed. When the key is fetched from localstorage, it fetches and renders the pointer key's value. These lines are also present in the render function but since we need async calls, the lines are also partially copied to the componentDidUpdated method.
Previously, the component wasn't using states but I store the cell data, once fetched from pointer key, to state upon the resolution of promise callbacks and the component's DidMount function updates the cell value accordingly.

Reply:
> These lines are also present in the render function 

Can you post a link to that render function that contains these lines?

Reply:
Before it used to simply get the cell values from props and no async calls were place.
https://github.com/parse-community/parse-dashboard/blob/1ddf29e84404ed36db86fc496fab82145add8bb1/src/components/BrowserCell/BrowserCell.react.js#L217

I updated it to store cell values in state so when the pointer key column name is fetched from localstorage and the value ,for this column, is fetched using js SDK .get method, the component state is updated with the value.
https://github.com/parse-community/parse-dashboard/pull/1723#discussion_r690610276

The lines in the render function that check the data type and show how the cell should be rendered are present in render function (I wanted to clean it up too by moving the lines to a separate function and call it in the didMount and didUpdate component lifecycle function function but didn't get around to it)

These lines check the cell data type and how it should be rendered (I needed to re-render the cell data once it's fetched from the localStorage+SDK so I had to copy them to didUpdate function for when the cell data in the component state updates).
This is where the logic of how the cell should renders starts.
https://github.com/parse-community/parse-dashboard/blob/1ddf29e84404ed36db86fc496fab82145add8bb1/src/components/BrowserCell/BrowserCell.react.js#L222

and ends around this line
https://github.com/parse-community/parse-dashboard/blob/1ddf29e84404ed36db86fc496fab82145add8bb1/src/components/BrowserCell/BrowserCell.react.js#L330

Reply:
> I wanted to clean it up too by moving the lines to a separate function and call it in the didMount and didUpdate component lifecycle function function but didn't get around to it

Thanks for explaining this in detail, I understand now. I think it would be really great if you could do the clean-up before we merge this PR. Duplicate code is often difficult to maintain, if someone later makes a change only in 1 part of code, but not in the other.